pkg/filewatcher/filewatcher.go

```diff
@@ -0,0 +1,253 @@
+			events: make(chan fsnotify.Event, 1),
+			md5Sum: md5Sum,
+		}
+
+		return nil
+	}
+
+	// Build a new worker if not exist.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	if err = watcher.Add(parentPath); err != nil {
+		return err
+	}
+
+	md5Sum, err := getMd5Sum(cleanedPath)
+	if err != nil {
+		return fmt.Errorf("failed to get md5 sum for %s: %v", path, err)
+	}
+	wk := &worker{
+		watcher: watcher,
+		watchedFiles: map[string]*fileTracker{
+			cleanedPath: {
+				events: make(chan fsnotify.Event, 1),
+				md5Sum: md5Sum,
+			},
+		},
+	}
+	w.workers[parentPath] = wk
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok { // 'Events' channel is closed
+					return
+				}
+
+				for path, tracker := range wk.watchedFiles {
+					newSum, _ := getMd5Sum(path)
+					if newSum != "" && newSum != tracker.md5Sum {
+						tracker.md5Sum = newSum
+						tracker.events <- event
+						break
+					}
+				}
+			case <-watcher.Errors:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Remove is implementation of Remove interface of FileWatcher.
+func (w *fsNotifyWatcher) Remove(path string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	return w.remove(path)
+}
+
+func (w *fsNotifyWatcher) remove(path string) error {
+	cleanedPath, parentPath := formatPath(path)
+
+	worker, workerExist := w.workers[parentPath]
+	if !workerExist {
+		return fmt.Errorf("path %s already removed", path)
+	}
+
+	tracker, pathExist := worker.watchedFiles[cleanedPath]
+	if !pathExist {
+		return fmt.Errorf("path %s already removed", path)
+	}
+
+	delete(worker.watchedFiles, cleanedPath)
+	close(tracker.events)
+	if len(worker.watchedFiles) == 0 {
+		// Remove the watch if all of its paths have been removed.
+		delete(w.workers, parentPath)
+		return worker.watcher.Close()
+	}
+
+	return nil
+}
+
+// Close is implementation of Close interface of FileWatcher.
+func (w *fsNotifyWatcher) Close() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	var errors *multierror.Error
+	for _, worker := range w.workers {
+		for path := range worker.watchedFiles {
+			errors = multierror.Append(errors, w.remove(path))
+		}
+	}
+
+	return errors.ErrorOrNil()
+}
+
+// Events is implementation of Events interface of FileWatcher.
+func (w *fsNotifyWatcher) Events(path string) chan fsnotify.Event {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	cleanedPath, parentPath := formatPath(path)
+	worker, workerExist := w.workers[parentPath]
+	if !workerExist {
+		return nil
+	}
+
+	tracker, pathExist := worker.watchedFiles[cleanedPath]
+	if !pathExist {
+		return nil
+	}
+
+	return tracker.events
+}
+
+// Errors is implementation of Errors interface of FileWatcher.
+func (w *fsNotifyWatcher) Errors(path string) chan error {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	_, parentPath := formatPath(path)
+	worker, workerExist := w.workers[parentPath]
+	if !workerExist {
+		return nil
+	}
+
+	return worker.watcher.Errors
```

Comment:
Isn't there an impedance mismatch problem here? It looks like you're multiplexing the change events to multiple channels, so Events() method works ok as long as there is a single caller for a given file.

However, calling Errors with different file paths will cause the same channel will be returned. Errors will not be routed, or multiplexed correctly. Multiple callers will get different events randomly. Am I missing something?

Reply:
Oh,absolutely correctly.
I had a follow-up commit to cover this, but the commit did not got pushed to this branch as expected for some reason :(